### PR TITLE
Move enforcement of "no static type as type argument" to constraint validation

### LINKED ISSUE #345: Roslyn fails to diagnose static type as type argument, reports "Failed to emit module".
The following program should fail to compile (a static type is used as a type argument), but Roslyn does not diagnose the problem during binding and fails in emit with a useless diagnostic.

```
error CS7038: Failed to emit module 'CSharpSample1'.
```

```cs
using System;
using System.Linq;
using System.Runtime.CompilerServices;

class Program
{
    static void Main(string[] args)
    {
        M(default(C));
    }
    public static void M<T>(T t)
    {
    }
}

static class C
{
}
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #345 and #511

### CODE REVIEW

**Comment by gafter** [comment]
To: @AlekseyTs @agocke @VSadov @VladimirReshetnikov @jaredpar 

Can a couple of you please take a look?

**Comment by jaredpar** [comment]
:+1:

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs [hunk 1]
```diff
@@ -19113,6 +19118,10 @@ public class Test
 
             comp4.GetDiagnostics().Verify(expected);
         }
+
+        public void InferredStaticTypeArgument()
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: Oops, missing the test.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: [Fact]?

**Comment by gafter** [comment]
To: @AlekseyTs @agocke @VSadov @VladimirReshetnikov 

Thank you @jaredpar. Can I please have one more "Heck Yeah"?

**Comment by gafter** [comment]
To: @AlekseyTs @agocke @VSadov @VladimirReshetnikov

Can I please have one more "Heck Yeah"?

**Comment by jaredpar** [comment]
image

;)

**Comment by AlekseyTs** [comment]
LGTM

**Pull request merged**